python_modules/dagster/dagster/serdes/serdes.py

```diff
@@ -115,18 +119,28 @@ def whitelist_for_serdes(serializer):
 
     """
 
-    if isclass(serializer) and not issubclass(serializer, Serializer):
-        return _whitelist_for_serdes(whitelist_map=_WHITELIST_MAP, serializer=None)(serializer)
+    if __cls is not None:
+        check.class_param(__cls, "__cls")
+        return _whitelist_for_serdes(whitelist_map=_WHITELIST_MAP, serializer=None)(__cls)
```

Comment:
<!-- GRAPHITE_METADATA {"trueLocation":{"startLine":105,"endLine":124}} -->
<i class='graphite__hidden'>[Re: lines 105 to 124]<br/><br/></i>whats going on here? if the tests are passing i assume it works but its not obvious to me whats going on here. Might be worth leaving a comment in code if there is something clever at play

Reply:
We are distinguishing between the two use cases of the decorator-- with args and without args. This is a real pain for type-checkers because the logic of the decorator is totally different depending on which one you're using. Either:

- the first arg is a decorator param, and you have to return another decorator
- the first arg is the decorated function, and you return the decorated function

The best way to distinguish these cases is to use only named arguments for decorator parameters. I changed `whitelist_for_serdes` to enforce this using the single asterisk in the signature. It was already true of all-but-one callsite anyway (wherever the `serializer` parameter was passed, it was named).

IMO the logic is now much more straightforward. Previously the first arg was named `serializer`, and we were checking to see whether it was _actually_ a serializer (parameter case) or not (decorated class case). Now these two are no longer assigned to the same name `serializer`. `serializer` must be passed as a keyword arg, and is always actually a serializer. `__cls` is always the decorated class (`None` in the case we called the decorator with params).

Do you still think I need to add a comment explaining this?

BTW, since Dagster makes such heavy use of decorators that take params, IMO this pattern should be implemented throughout the codebase (decorator params are always keywords). Do we have any legitimate use case for passing positional args as decorator params? I remember when I started with Dagster, this bit me, pyright immediately starting yelling at me about `@solid` bc it got confused between its direct and parametrized invocations.

Reply:
Cool thanks for explaining - having this written in the PR provides something if someone goes digging.

I think this is the first time I have seen `, *,` used to force keyword args. 

> I've added a brief comment.

This is probably sufficient. Maybe some type of "force extra params to be kwargs" comment to explain thats whats happening but its unclear the best place to stash that.

Reply:
> Do we have any legitimate use case for passing positional args as decorator params?

Depending on definition of legitimate, there are likely very few cases if any where keywords are not the clear way to go. 

I support changing any of our internal decorators to this stricter pattern, but we have to be very thoughtful about changing user facing APIs which could break people.